### ISSUE TITLE
tensorrt test failed

### ISSUE THREAD

**Issue by twmht** [open-mmlab/mmcv#1454]
when trying to run test on test_tensorrt.py, there are many cuda errors.

My environment is cuda11+pytorch1.8.0+tensorrt7.1.3.4

I found out when forwarding  pytorch module with cpu mode, the error could be gone.

or remove the tensorrt part, the error could be gone.

so the issue is that when forwarding pytorch module with gpu mode and run tensorrt module together, the cuda error may be throw.

Here are the test log

[log.txt](https://github.com/open-mmlab/mmcv/files/7465778/log.txt)

**Comment by zhouzaida** [comment]
hi @twmht, the log file you provided did not include cuda error information.
```
Traceback (most recent call last):
  File "tools/rapid.py", line 4, in <module>
    import torch
  File "/home/acer/.pyenv/versions/pytorch-1.8/lib/python3.7/site-packages/torch/__init__.py", line 196, in <module>
    from torch._C import *
RuntimeError: KeyboardInterrupt: 
```

**Comment by twmht** [comment]
oops.

I may upload the wrong file. I have updated the post and please download the file again.

**Comment by zhouzaida** [comment]
Got it. Please @grimoire have a look.

**Comment by grimoire** [comment]
I am not 100% sure, but I guess it is caused by memory conflict.
Both PyTorch and TensorRT have their own memory pool, which might pollute each other when you are using both of them.
It is recommended to use TensorRT only in the real environment, or you can create a [GPU allocator](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/GpuAllocator.html) to feed pytorch memory to TensorRT.

**Comment by twmht** [comment]
@grimoire 

i don't understand the memory pool, but why we have memory conflict if they have their "own" memory pool?

**Comment by grimoire** [comment]
Update:

I find something that might be related: https://github.com/pytorch/pytorch/issues/52663
I did use thrust in nms ops, which might use cub implementation `_ZN6thrust8cuda_cub11sort_by_keyINS0_17execute_on_streamEPfPiNS_7greaterIfEEEEvRNS0_16execution_policyIT_EET0_SB_T1_T2_`. And if I swap the order in `test_batched_nms`, the TensorRT nms ops crashed after  `thrust::sort_by_key`
According to this issue, cub might have a conflict with torch.

**Comment by twmht** [comment]
@grimoire 

what cause the conflict? in my test log, there are many test failed, not only test_batched_nms.

**Comment by grimoire** [comment]
Just comment test_batched_nms and all other test will be passed.

**Comment by twmht** [comment]
Interested. Why is pytorch conflict with thrust::sort_by_key?

**Comment by grimoire** [comment]
Both PyTorch and thrust use [cub](https://github.com/NVIDIA/cub) in their source code. And there is a `template function` with `static variable` https://github.com/NVIDIA/cub/blob/499a7bad3416fcc71a7c50351d6b3cdbf3fbbc27/cub/util_device.cuh#L210.

After compile and load both libraries, they have two different `DeviceCountCachedValue` with same static variable `cache`. That's where the problem comes from. According to issures in thrust and PyTorch, add compile flags `-Xcompiler -fno-gnu-unique` might solve the problem.

reference:
- https://github.com/NVIDIA/thrust/issues/1401#issuecomment-806403746

**Comment by twmht** [comment]
@grimoire 

thank you! Great explain!

But when running mmcv nms ops alone without running tensorrt module, the test would be fine.  Why tensorrt introduce the influence?

**Comment by grimoire** [comment]
[TensorRT nms](https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/tensorrt/plugins/trt_nms_kernel.cu) and [MMCV nms](https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/pytorch/cuda/nms_cuda.cu) are different. In MMCV implementation, we use PyTorch to do the sort or topk, which will not bring another `cub`.

The error is caused by the way how compiler process static variable in the template. Here is an example (from a blog)
[unique_bind.zip](https://github.com/open-mmlab/mmcv/files/7480237/unique_bind.zip)
Try comment or uncomment `SET(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -fno-gnu-unique")` in CMakeLists.txt. Recompile the project and you will have different output.

I am using torch1.8.1 and the error still exists. Not sure if they have a fix in 1.10.0.
